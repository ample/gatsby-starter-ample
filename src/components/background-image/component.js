import React from "react"
import PropTypes from "prop-types"
import GatsbyBackgroundImage from "gatsby-background-image"
import classNames from "classnames"

import styles from "./styles.module.scss"

const BackgroundImage = ({ children, className, image }) => {
  // Combine wrapping class names.
  const wrapperClassNames = classNames(className, styles.background_image)
  // If there is no image, render a wrapper and the children.
  if (!image) return <div className={wrapperClassNames}>{children}</div>
  // Should return either fluid or fixed if setup properly.
  const imageType = Object.keys(image.childImageSharp)[0]
  // Set props for the wrapping element.
  let props = {
    className: wrapperClassNames,
    [imageType]: image.childImageSharp[imageType]
  }
  // Render the content.
  return <GatsbyBackgroundImage {...props}>{children}</GatsbyBackgroundImage>
}

BackgroundImage.propTypes = {
  /**
   * Content to render inside the wrapping element (the one with the background
   * image attached).
   */
  children: PropTypes.node,
  /**
   * Additional classes on the wrapping element (the one with the background
   * image).
   */
  className: PropTypes.string,
  /**
   * Object representing the image values to pass on to gatsby-background-image.
   */
  image: PropTypes.object
}

BackgroundImage.defaultProps = {}

export default BackgroundImage
