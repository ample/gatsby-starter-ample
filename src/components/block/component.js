import React from "react"
import PropTypes from "prop-types"

import componentMap from "./component-map"

const Block = props => {
  // If no data was passed, render nothing.
  if (!props) return null

  // Name of the component to render comes from the componentMap, defined in
  // component-map.js.
  const TagName = componentMap[props.template]

  // If the tag is not found, throw a console error and exit.
  if (!TagName) {
    console.error(`${props.template} mapping is not supported.`)
    return null
  }

  // Render the component
  return <TagName {...props} />
}

/**
 * This is open-ended. All other props pass-through this component directly to
 * the component.
 */
Block.propTypes = {
  /**
   * The name of the component, which gets mapped to the actual component in
   * component-map.js.
   */
  template: PropTypes.string
}

Block.defaultProps = {}

export default Block
