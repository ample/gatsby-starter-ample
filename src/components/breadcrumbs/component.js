import React from "react"
import PropTypes from "prop-types"
import classNames from "classnames"

import Link from "@src/components/link"
import SVG from "@src/components/svg"

import styles from "./styles.module.scss"

const Breadcrumbs = ({ className, links, separator }) => (
  <ul className={classNames(styles.breadcrumbs, { [className]: className })}>
    {links.map(({ label, icon, url }, index) => (
      <li
        className={classNames({ [styles[`only_icon`]]: icon })}
        key={index}
        data-separator={separator}
      >
        <Link className={styles.crumb} to={url} key={index}>
          <>
            {icon && <SVG name={icon} />}
            {label && label}
          </>
        </Link>
      </li>
    ))}
  </ul>
)

Breadcrumbs.propTypes = {
  /**
   * Additional classes on the wrapping element.
   */
  className: PropTypes.string,
  /**
   * An array of links used to build the breadcrumbs
   * If a label and icon exist, the icon will be added to the left of the label.
   */
  links: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      icon: PropTypes.string,
      url: PropTypes.string
    })
  ),
  /**
   * The character used to separate the breadcrumbs.
   */
  separator: PropTypes.string
}

Breadcrumbs.defaultProps = {
  separator: "/"
}

export default Breadcrumbs
