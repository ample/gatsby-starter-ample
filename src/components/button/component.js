import React from "react"
import PropTypes from "prop-types"
import classNames from "classnames"

import Link from "../link"
import SVG from "../svg"

import styles from "./styles.module.scss"

const Button = ({ children, className, onClick, theme, to }) => {
  const classes = classNames(styles.button, {
    [className]: className,
    [styles[`theme_${theme}`]]: theme
  })

  return (
    <Link to={to} className={classes} onClick={onClick}>
      {children}
      {theme === "arrow" && <SVG name="angle-right" />}
    </Link>
  )
}

Button.propTypes = {
  /**
   * Additional classes on the link element.
   */
  className: PropTypes.string,
  /**
   * Text rendered to the screen inside the button.
   */
  children: PropTypes.string.isRequired,
  /**
   * Specifies a click Function and uses the button element
   */
  onClick: PropTypes.func,
  /**
   * Specifies the theme of button
   */
  theme: PropTypes.oneOf(["arrow", "default", "outline"]),
  /**
   * The href attribute for the link rendered to the screen.
   */
  to: PropTypes.string.isRequired
}

Button.defaultProps = {
  theme: "default"
}

export default Button
