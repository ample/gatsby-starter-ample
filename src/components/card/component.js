import React from "react"
import PropTypes from "prop-types"
import classNames from "classnames"

import Link from "@src/components/link"
import Button from "@src/components/button"
import Image from "@src/components/image"

import styles from "./styles.module.scss"

const Card = ({ body, button, heading, image, textAlignment, theme, url }) => {
  const classes = classNames(styles.card, {
    [styles[`text_alignment_${textAlignment}`]]: textAlignment,
    [styles[`theme_${theme}`]]: theme
  })

  let card = (
    <div className={classes}>
      {image && <Image className={styles.card_image} src={image} />}
      {heading && <h3>{heading}</h3>}
      {body && <div dangerouslySetInnerHTML={{ __html: body }} />}
      {button && <Button {...button} />}
    </div>
  )

  if (url) {
    card = <Link to={url}>{card}</Link>
  }

  return card
}

Card.propTypes = {
  /**
   * Specifies the content
   * that is displayed under the heading
   */
  body: PropTypes.string,
  /**
   * Specifies the label and url
   * if the card has a button
   */
  button: PropTypes.exact({
    label: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
  }),
  /**
   * Specifies the heading
   */
  heading: PropTypes.string,
  /**
   * Specifies the image src and alt text
   */
  image: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  /**
   * Specifies the alignment of the card's content
   */
  textAlignment: PropTypes.oneOf(["left", "center", "right"]),
  /**
   * Specifies the card theme
   */
  theme: PropTypes.oneOf(["graphic", "image"]),
  /**
   * Specifies the url for the card
   *
   * NOTE: Only use this prop,
   * if the card DOES NOT have a button
   * This will make the entire card a link
   */
  url: PropTypes.string
}

Card.defaultProps = {
  textAlignment: "left"
}

export default Card
