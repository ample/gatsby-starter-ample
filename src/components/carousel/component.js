import React from "react"
import PropTypes from "prop-types"
import SwiperCore, { A11y, Navigation } from "swiper"
import { Swiper, SwiperSlide } from "swiper/react"

import "swiper/swiper.scss"
import "swiper/components/navigation/navigation.scss"

// import styles from "./styles.module.scss"

SwiperCore.use([Navigation, A11y])

const Carousel = ({ slides }) => {
  return (
    <>
      <div>
        <button id="prev-slide">Back</button>
        <button id="next-slide">Next</button>
      </div>
      <Swiper
        spaceBetween={0}
        slidesPerView={1}
        navigation={{ nextEl: "#next-slide", prevEl: "#prev-slide" }}
        a11y
      >
        {slides.map((slide, idx) => (
          <SwiperSlide key={idx}>{slide.body}</SwiperSlide>
        ))}
      </Swiper>
    </>
  )
}

Carousel.propTypes = {
  /**
   * An array of data used to render the carousel's slides.
   */
  slides: PropTypes.arrayOf(
    PropTypes.shape({
      body: PropTypes.string.isRequired
    })
  ).isRequired
}

Carousel.defaultProps = {}

export default Carousel
