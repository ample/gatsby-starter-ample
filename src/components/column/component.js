import React from "react"
import PropTypes from "prop-types"
import classNames from "classnames"

import styles from "./styles.module.scss"

import Block from "../block"

const Column = ({ className, blocks, config }) => {
  const wrapperClasses = classNames(styles.column, {
    [className]: className,
    [`mb-${config.margin_bottom}`]: config.margin_bottom,
    [`text-${config.text_align}`]: config.text_align,
    [styles[`col_width_${config.width}`]]: config.width
  })

  return (
    <div className={wrapperClasses}>
      {blocks.map((block, idx) => (
        <Block key={idx} {...block} />
      ))}
    </div>
  )
}

Column.propTypes = {
  /**
   * A CSS class adding to the wrapping element.
   */
  className: PropTypes.string,
  /**
   * An array of component objects that get passed on to the <Block />.
   */
  blocks: PropTypes.arrayOf(PropTypes.object),
  /**
   * An object that controls the styling of the container on screen.
   */
  config: PropTypes.shape({
    margin_bottom: PropTypes.string,
    text_align: PropTypes.oneOf(["left", "center", "right", ""]),
    width: PropTypes.oneOf(["full", "1/4", "1/3", "1/2", "2/3", "3/4", ""])
  })
}

Column.defaultProps = {
  blocks: [],
  config: {}
}

export default Column
