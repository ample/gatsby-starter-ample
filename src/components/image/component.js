import React from "react"
import PropTypes from "prop-types"
import Img from "gatsby-image"
import classNames from "classnames"
import dig from "object-dig"
import startCase from "lodash/startCase"
import path from "path"

import styles from "./styles.module.scss"

export const defaultAltAttribute = image => {
  const filename = path.basename(image, path.extname(image))
  return startCase(filename)
}

const Image = ({ alt, className, src, ...props }) => {
  const classes = classNames(styles.image, { [className]: className })

  // ---------------------------------------- | Gastby Image

  if (dig(src, "childImageSharp", "fluid") || dig(src, "childImageSharp", "fixed")) {
    let imageName = dig(src, "childImageSharp", "fluid", "src")

    if (dig(src, "childImageSharp", "fixed")) {
      imageName = dig(src, "childImageSharp", "fixed", "src")
    }

    return (
      <Img
        className={classes}
        alt={alt || defaultAltAttribute(imageName)}
        {...src.childImageSharp}
      />
    )
  }

  // ---------------------------------------- | Native Image

  if (typeof src === "string") {
    return <img className={classes} src={src} alt={alt || defaultAltAttribute(src)} {...props} />
  }

  // ---------------------------------------- | Invalid src

  return null
}

Image.propTypes = {
  /**
   * Specifies the image alt attribute.
   */
  alt: PropTypes.string,
  /**
   * Classes attributed to the wrapping element.
   */
  className: PropTypes.string,
  /**
   * Specifies the image src attribute as a string or Gatsby fixed/fluid image
   * object ([Gatsby Docs](https://www.gatsbyjs.org/packages/gatsby-image/)).
   */
  src: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  /**
   * Support for directly passing styles to the image component.
   */
  styles: PropTypes.object
}

Image.defaultProps = {}

export default Image
