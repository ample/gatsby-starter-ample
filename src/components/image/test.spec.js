import React from "react"
import renderer from "react-test-renderer"

import Image, { fixtures, transformer } from "."
import { defaultAltAttribute } from "./component"

// ---------------------------------------- | Controller

describe("Image Controller", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Image {...fixtures.native} />).toJSON()
    expect(tree.props.src).toEqual(fixtures.native.src)
  })
  it("transforms props", () => {
    const tree = renderer.create(<Image image={fixtures.native.src} />).toJSON()
    expect(tree.props.src).toEqual(fixtures.native.src)
  })
})

// ---------------------------------------- | Component

describe("Image", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<Image src="https://via.placeholder.com/300" alt="Image from https://placehold.it" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("will render null if the src object is not the correct format", () => {
    const src = { childImageSharp: { wrongKey: { src: "..." } } }
    const tree = renderer.create(<Image src={src} alt="alt ..." />).toJSON()
    expect(tree).toBeNull()
  })
  it("will render with a proper fluid image", () => {
    const tree = renderer.create(<Image src={fixtures.fluid} alt="alt ..." />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("will render with a proper fluid image and a default alt attribute", () => {
    const tree = renderer.create(<Image src={fixtures.fluid} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("will render with a proper fixed image", () => {
    const tree = renderer.create(<Image src={fixtures.fixed} alt="alt ..." />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

// ---------------------------------------- | Alt Attribute

describe("defaultAltAttribute", () => {
  it("creates a humanized version of a URL string", () => {
    const filename = "/uploads/image-01.jpg"
    expect(defaultAltAttribute(filename)).toEqual("Image 01")
  })
})

// ---------------------------------------- | Transformer

describe("Image Transformer", () => {
  it("sets children from label", () => {
    const result = transformer({ image: "/hello-world.jpg" })
    expect(result.src).toEqual("/hello-world.jpg")
  })
  it("keeps the original props", () => {
    const result = transformer({ image: "/hello-world.jpg" })
    expect(result.image).toEqual("/hello-world.jpg")
  })
  it("brings along other props", () => {
    const result = transformer({ image: "/hello-world.jpg", hello: "world" })
    expect(result.hello).toEqual("world")
  })
  it("gives precedence to the correct props", () => {
    const result = transformer({ image: "/hello-world.jpg", src: "/real-image.png" })
    expect(result.src).toEqual("/real-image.png")
  })
})
