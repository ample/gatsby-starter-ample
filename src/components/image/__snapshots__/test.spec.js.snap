// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders correctly 1`] = `
<img
  alt="Image from https://placehold.it"
  className="image"
  src="https://via.placeholder.com/300"
/>
`;

exports[`Image will render with a proper fixed image 1`] = `null`;

exports[`Image will render with a proper fluid image 1`] = `null`;

exports[`Image will render with a proper fluid image and a default alt attribute 1`] = `null`;
